sydent/http/servlets/__init__.py

```diff
@@ -163,45 +163,32 @@ def inner(self, request: Request, *args, **kwargs) -> bytes:
 
 
 def asyncjsonwrap(f):
-    def reqDone(resp: Dict[str, Any], request: Request) -> None:
-        """
-        Converts the given response content into JSON and encodes it to bytes, then
-        writes it as the response to the given request with the right headers.
-
-        :param resp: The response content to convert to JSON and encode.
-        :param request: The request to respond to.
-        """
-        request.setHeader("Content-Type", "application/json")
-        request.write(dict_to_json_bytes(resp))
-        request.finish()
-
-    def reqErr(failure: Failure, request: Request) -> None:
-        """
-        Logs the given failure. If the failure is a MatrixRestError, writes a response
-        using the info it contains, otherwise responds with 500 Internal Server Error.
-
-        :param failure: The failure to process.
-        :param request: The request to respond to.
-        """
-        request.setHeader("Content-Type", "application/json")
-        if failure.check(MatrixRestError) is not None:
-            request.setResponseCode(failure.value.httpStatus)
-            request.write(
-                dict_to_json_bytes(
-                    {"errcode": failure.value.errcode, "error": failure.value.error}
+    async def render(f, *args, **kwargs):
+        request = args[1]
+        try:
+            result = await f(*args, **kwargs)
+            request.setHeader("Content-Type", "application/json")
+            request.write(dict_to_json_bytes(result))
+            request.finish()
+        except Exception:
+            f = failure.Failure()
+            request.setHeader("Content-Type", "application/json")
+            if f.check(MatrixRestError) is not None:
```

Comment:
Instead of returning the error into a failure and then pulling the exception back out (`f.value` is the exception`) I think it would make more sense to `except Exception as e` and then we can do `if isinstance(e, MatrixRestError)` to see if it is the type we expect.

To further clarify this it seems that we're *always* returning the same `Content-Type` header, so we can move that higher up in the function (before calling `f`) and then catch specific exceptions instead, something like:

```python
request.setHeader("Content-Type", ...)
try:
    response = await f(...)
    request.write/finish
except MatrixRestError as e:
    request.setResponseCode(e.httpStatus)
    request.write(...)
except:
    ...
    request.setResponseCode(500)
    request.write(...)
```

Reply:
I added a `failure` on the last `except` stanza so that this line: `logger.error("Request processing failed: %r, %s", f, f.getTraceback())` would still work, I hope that's okay!

Reply:
Ah, I had missed that we used a failure there. That should work OK!